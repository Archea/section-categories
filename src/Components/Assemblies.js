import React, { Component } from 'react'
import AssemblyList from './AssemblyList'
import AssemblyCrumb from './AssemblyCrumb'
import ProductFamilyListA from './ProductFamilyListA'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
/*
  an emptly assembly object this is the only state of
  the assembly component tree
*/
const initialState = {
  selected: {
    id: '',
    description: '',
    code: '',
    level: 0
  }
}

export class Assemblies extends Component {
  state = initialState
  //using local state for brevity.
  setSelected = selection => {
    this.setState(state => {
      return { ...state, selected: selection }
    })
  }
  /*
    this decomposes the current crumb by level
    the level of the button clicked is sliced down to get
    higher up the current assembly three
    the code that matches is the new selection we want.
  */
  selectCrumb = level => {
    if (level === 0) this.setSelected(initialState)
    const code = this.state.selected.code.slice(0, CodeSlice(level))
    const setTo = this.props.getAll.allUniformatClassifications.find(
      a => a.code === code
    )
    this.setSelected(setTo)
  }

  render() {
    if (this.props.getAll && this.props.getAll.loading) {
      return <div>Loading</div>
    }

    if (this.props.getAll && this.props.getAll.error) {
      console.log('Get All Assemblies error: ' + this.props.getAll.error)
      return <div>Error</div>
    }
    //This should always return, the content service is down if it fails
    const assemblies = this.props.getAll.allUniformatClassifications
    if (!assemblies) {
      return <div>Error no assemblies !!</div>
    }
    //Don't make a list with an empty id, aka no selection, avoid call and render
    const productFamilies =
      this.state.selected.id === '' ? (
        <div>none selected</div>
      ) : (
        <ProductFamilyListA assemblieId={this.state.selected.id} />
      )
    //clearing the selection effectively gets you back to the top of the tree
    return (
      <div className="Assemblies">
        <button onClick={e => this.setSelected(initialState.selected)}>
          Top
        </button>
        <AssemblyCrumb
          selected={this.state.selected}
          onCrumb={this.selectCrumb}
        />
        <AssemblyList
          assemblies={assemblies}
          selected={this.state.selected}
          select={this.setSelected}
        />
        {productFamilies}
      </div>
    )
  }
}
/*
  Assembly codes follow a strict schema like mf section#s
  we can use that to decompse them by charactar position.
  This returns the correct character position for a given level
  in a code.
*/
const CodeSlice = level => {
  switch (level) {
    case 1:
      return 1
    case 2:
      return 3
    case 3:
      return 5
    case 4:
      return 8
    case 5:
      return 10
    default:
      return 0
  }
}

//this is the graphql query to get all the assemblies
const ASSEMBLIES_QUERY = gql`
  query GetAll {
    allUniformatClassifications {
      id
      description
      code
      level
    }
  }
`
export default graphql(ASSEMBLIES_QUERY, { name: 'getAll' })(Assemblies)
