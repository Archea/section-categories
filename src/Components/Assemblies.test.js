import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, configure } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { Assemblies } from './Assemblies'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

describe('Assemblies', () => {
  const mockAssemblies = [
    {
      id: '1',
      name: 'bar',
      prefix: 'fixed',
      description: 'a solid steele cylnder',
      category: 'structural restraints',
      productFamilies: ['tough things', 'cage stuff']
    },
    {
      id: '2',
      name: 'foosle',
      prefix: 'fix',
      description: 'a stone cylnder',
      category: 'structural supports',
      productFamilies: ['tough things', 'hard stuff']
    }
  ]
  const mockGetAll = jest.fn
  const mockprops = {
    getAll: {
      allUniformatClassifications: mockAssemblies
    }
  }
  test('it at least shallow renders and match snapshot', () => {
    const wrapper = shallow(<Assemblies {...mockprops} />)
    wrapper.setProps(mockprops)
    expect(toJSON(wrapper)).toMatchSnapshot()
  })
})
