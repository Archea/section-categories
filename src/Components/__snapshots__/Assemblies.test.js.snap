// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assemblies it at least shallow renders and match snapshot 1`] = `
<div
  className="Assemblies"
>
  <button
    onClick={[Function]}
  >
    Top
  </button>
  <AssemblyCrumb
    onCrumb={[Function]}
    selected={
      Object {
        "code": "",
        "description": "",
        "id": "",
        "level": 0,
      }
    }
  />
  <AssemblyList
    assemblies={
      Array [
        Object {
          "code": "A",
          "description": "SUBSTRUCTURE",
          "id": "63ef3b02-faa5-44f5-a677-99775fdaeffb",
          "level": 1,
        },
        Object {
          "code": "A10",
          "description": "FOUNDATIONS",
          "id": "7777f4e9-bc0d-45cf-b66d-b97ad4638dc0",
          "level": 2,
        },
      ]
    }
    select={[Function]}
    selected={
      Object {
        "code": "",
        "description": "",
        "id": "",
        "level": 0,
      }
    }
  />
  <div>
    none selected
  </div>
</div>
`;
